### PR TITLE
(#7) Added Redis cluster support

### PR DESCRIPTION
**Pull request by s1lver**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #7

- Changed CI to run single and clustered Redis
- Added identical tests with a connection to the Redis cluster
- Updated predis to 2.0 (#4)
- Updated predis to 2.1 (#12)
- Added base support Redis cluster

### CODE REVIEW

File: src/RedisCache.php [hunk 1]
```diff
@@ -32,21 +36,63 @@ final class RedisCache implements CacheInterface
...
+     */
+    public function isCluster(): bool
+    {
+        $isCluster = false;
+        /** @psalm-suppress MixedAssignment, PossibleRawObjectIteration */
+        foreach ($this->connections as $connection) {
+            /** @var StreamConnection $connection */
+            $info = (string)$connection->executeCommand(new RawCommand('INFO', ['Cluster']));
+            $clusterEnabled = explode('cluster_enabled:', trim($info));
+            $isCluster = (bool)$clusterEnabled[1];
```
Comment (samdark) [inline src/RedisCache.php]: Can we return true here immediately?

**Review by samdark — APPROVED** [review]
Overall looks good.

File: src/RedisCache.php (hunk 1 above)
Comment (s1lver) [reply src/RedisCache.php]: Fixed

**Comment by samdark** [comment]
Ready for review/merge?

**Comment by s1lver** [comment]
> Ready for review/merge?

It is also necessary to implement work with **MGET** and **MSET** 
... and fixed Scrutinizer

**Comment by xepozz** [comment]
Don't worry about Scrutinizer

File: src/RedisCache.php [hunk 2]
```diff
@@ -27,20 +30,64 @@
  */
 final class RedisCache implements CacheInterface
 {
+    /**
+     * @var array<ConnectionInterface>|ConnectionInterface $connections Predis connections instance to use
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
     * @var array<ConnectionInterface>|ConnectionInterface $connections Predis connection instances to use.
```

File: src/RedisCache.php [hunk 3]
```diff
@@ -27,20 +30,64 @@
...
     /**
      * @param ClientInterface $client Predis client instance to use.
      */
     public function __construct(private ClientInterface $client)
     {
+        $this->connections = $this->client->getConnection();
     }
 
+    /**
+     * Checking Predis cluster usage
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
     * Returns whether Predis cluster is used.
```

File: src/RedisCache.php [hunk 4]
```diff
@@ -27,20 +30,64 @@
...
+     *
+     * @return bool
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
     * @return bool Whether Predis cluster is used.
```

File: src/RedisCache.php [hunk 5]
```diff
@@ -27,20 +30,64 @@
...
+    }
+
+    /**
+     * @param string $key
+     * @param mixed|null $default
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return mixed
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 6]
```diff
@@ -27,20 +30,64 @@
...
 
+    /**
+     * @param string $key
+     * @param mixed $value
+     * @param DateInterval|int|null $ttl
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
    /**
     * @param string $key
     * @param mixed $value
     * @param DateInterval|int|null $ttl
     *
     * @throws InvalidArgumentException
     *
     * @return bool
     */
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 7]
```diff
@@ -54,42 +101,90 @@ public function set(string $key, mixed $value, null|int|DateInterval $ttl = null
...
         return $result !== null;
     }
 
+    /**
+     * @param string $key
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 8]
```diff
@@ -27,20 +30,64 @@
...
     {
+        $this->connections = $this->client->getConnection();
     }
 
+    /**
+     * Checking Predis cluster usage
+     *
+     * @return bool
+     */
+    public function isCluster(): bool
```
Comment (samdark) [inline src/RedisCache.php]: Should it be `public`?

File: src/RedisCache.php [hunk 9]
```diff
@@ -54,42 +101,90 @@ public function set(string $key, mixed $value, null|int|DateInterval $ttl = null
...
     public function delete(string $key): bool
     {
         return !$this->has($key) || $this->client->del($key) === 1;
     }
 
+    /**
+     * If a cluster is used, all nodes will be cleared
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
    /**
     * @inheritDoc
     *
     * If a cluster is used, all nodes will be cleared.
     */
```

File: src/RedisCache.php [hunk 10]
```diff
@@ -54,42 +101,90 @@ public function set(string $key, mixed $value, null|int|DateInterval $ttl = null
...
     }
 
+    /**
+     * @param iterable<string> $keys
+     * @param mixed $default
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return iterable<string, mixed>
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 11]
```diff
@@ -54,42 +101,90 @@ public function set(string $key, mixed $value, null|int|DateInterval $ttl = null
...
     }
 
+    /**
+     * @param iterable $values
+     * @param DateInterval|int|null $ttl
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 12]
```diff
@@ -107,30 +202,38 @@ public function setMultiple(iterable $values, null|int|DateInterval $ttl = null)
...
+        return !in_array(null, (array)$results, true);
     }
 
+    /**
+     * @param iterable $keys
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

File: src/RedisCache.php [hunk 13]
```diff
@@ -142,12 +245,21 @@ public function deleteMultiple(iterable $keys): bool
...
         return empty($keys) || $this->client->del($keys) === count($keys);
     }
 
+    /**
+     * @param string $key
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return bool
+     */
```
Comment (samdark) [inline src/RedisCache.php]:
```suggestion
```

Should be inherited automatically from cache interface.

**Comment by samdark** [comment]
Also, it would be great to mention cluster support in README.

**Comment by samdark** [comment]
Code overall is 👍

**Comment by samdark** [comment]
👍

**Pull request merged**